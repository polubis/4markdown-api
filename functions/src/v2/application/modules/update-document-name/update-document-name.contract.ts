import { documentNameSchema } from '@utils/document-schemas';
import { type Date, validators } from '@utils/validators';
import { z } from 'zod';

const updateDocumentNamePayloadSchema = z.object({
  id: validators.id,
  mdate: validators.date,
  name: documentNameSchema,
});

type UpdateDocumentNamePayload = z.infer<
  typeof updateDocumentNamePayloadSchema
>;

type UpdateDocumentNameDto = {
  mdate: Date;
  name: string;
};

export type { UpdateDocumentNamePayload, UpdateDocumentNameDto };
export { updateDocumentNamePayloadSchema };
