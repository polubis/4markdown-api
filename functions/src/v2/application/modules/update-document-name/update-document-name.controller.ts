import { errors } from '../../utils/errors';
import { protectedController } from '../../utils/controller';
import { z } from 'zod';
import { type Date, validators } from '../../utils/validators';
import { parse } from '../../utils/parse';
import type {
  DocumentModel,
  DocumentsModel,
} from '../../../domain/models/document';
import { nowISO } from '../../../libs/helpers/stamps';
import { DBInstance } from '../../database/database';

const payloadSchema = z.object({
  id: validators.id,
  mdate: validators.date,
  name: validators.document.name,
});

type Payload = z.infer<typeof payloadSchema>;

type Dto = {
  mdate: Date;
};

const containsDuplicateInAccessibleDocuments = async (
  payload: Payload,
  db: DBInstance,
): Promise<boolean> => {
  const allDocumentsSnap = (await db.collection(`docs`).get()).docs;

  for (const userDocumentsSnap of allDocumentsSnap) {
    const userDocuments = userDocumentsSnap.data() as DocumentsModel;

    const hasDuplicate = Object.entries(userDocuments).some(
      ([documentId, document]) =>
        documentId !== payload.id &&
        document.visibility === `permanent` &&
        document.name === payload.name,
    );

    if (hasDuplicate) return true;
  }

  return false;
};

const updateDocumentNameController = protectedController<Dto>(
  async (rawPayload, { uid, db }) => {
    const userDocumentsRef = db.collection(`docs`).doc(uid);
    const [payload, userDocumentsSnap] = await Promise.all([
      parse(payloadSchema, rawPayload),
      userDocumentsRef.get(),
    ]);
    const userDocuments = userDocumentsSnap.data() as
      | DocumentsModel
      | undefined;

    if (!userDocuments) throw errors.notFound(`Document data not found`);

    const userDocument = userDocuments[payload.id] as DocumentModel | undefined;

    if (!userDocument) {
      throw errors.notFound(`Document not found`);
    }

    if (payload.mdate !== userDocument.mdate) {
      throw errors.outOfDate(`The document has been already changed`);
    }

    const hasDuplicateInOwnDocuments = Object.entries(userDocuments).some(
      ([id, document]) => id !== payload.id && document.name === payload.name,
    );

    if (
      hasDuplicateInOwnDocuments ||
      (userDocument.visibility === `permanent` &&
        (await containsDuplicateInAccessibleDocuments(payload, db)))
    ) {
      throw errors.exists(
        `Document with provided name already exists, please change name`,
      );
    }

    const mdate = nowISO();

    await userDocumentsRef.update(<DocumentsModel>{
      [payload.id]: {
        ...userDocument,
        name: payload.name,
        mdate,
      },
    });

    return { mdate };
  },
);

export { updateDocumentNameController };
