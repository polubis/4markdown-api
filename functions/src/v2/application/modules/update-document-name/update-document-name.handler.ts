import { nowISO } from '@libs/helpers/stamps';
import type { DocumentModel, DocumentsModel } from '@domain/models/document';
import { errors } from '@utils/errors';
import { type ProtectedControllerHandlerContext } from '@utils/controller';
import {
  type UpdateDocumentNameDto,
  type UpdateDocumentNamePayload,
} from './update-document-name.contract';
import { type DBInstance } from '@database/database';

const containsDuplicateInAccessibleDocuments = async (
  payload: UpdateDocumentNamePayload,
  db: DBInstance,
): Promise<boolean> => {
  const allDocumentsSnap = (await db.collection(`docs`).get()).docs;

  for (const userDocumentsSnap of allDocumentsSnap) {
    const userDocuments = userDocumentsSnap.data() as DocumentsModel;

    const hasDuplicate = Object.entries(userDocuments).some(
      ([documentId, document]) =>
        documentId !== payload.id &&
        document.visibility === `permanent` &&
        document.path === payload.name.path,
    );

    if (hasDuplicate) return true;
  }

  return false;
};

const updateDocumentNameHandler = async ({
  payload,
  context: { db, uid },
}: {
  payload: UpdateDocumentNamePayload;
  context: ProtectedControllerHandlerContext;
}): Promise<UpdateDocumentNameDto> => {
  const userDocumentsRef = db.collection(`docs`).doc(uid);
  const userDocumentsSnap = await userDocumentsRef.get();
  const userDocuments = userDocumentsSnap.data() as DocumentsModel | undefined;

  if (!userDocuments) throw errors.notFound(`Document data not found`);

  const userDocument = userDocuments[payload.id] as DocumentModel | undefined;

  if (!userDocument) {
    throw errors.notFound(`Document not found`);
  }

  if (payload.mdate !== userDocument.mdate) {
    throw errors.outOfDate(`The document has been already changed`);
  }

  const hasDuplicateInOwnDocuments = Object.entries(userDocuments).some(
    ([id, document]) =>
      id !== payload.id && document.path === payload.name.path,
  );

  if (
    hasDuplicateInOwnDocuments ||
    (userDocument.visibility === `permanent` &&
      (await containsDuplicateInAccessibleDocuments(payload, db)))
  ) {
    throw errors.exists(
      `Document with provided name already exists, please change name`,
    );
  }

  const mdate = nowISO();

  const model: DocumentsModel = {
    [payload.id]: {
      ...userDocument,
      path: payload.name.path,
      name: payload.name.raw,
      mdate,
    },
  };

  await userDocumentsRef.update(model);

  return { mdate };
};

export { updateDocumentNameHandler };
