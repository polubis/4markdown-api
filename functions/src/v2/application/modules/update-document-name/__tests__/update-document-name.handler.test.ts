import {
  DocumentModel,
  DocumentModelVisibility,
  DocumentsModel,
} from '@domain/models/document';
import { type UpdateDocumentNamePayload } from '../update-document-name.contract';
import { updateDocumentNameHandler } from '../update-document-name.handler';

const stampMock = `2024-12-01T00:00:00Z`;

jest.mock(`@libs/helpers/stamps`, () => ({
  nowISO: jest.fn(() => stampMock),
  uuid: jest.fn(() => `mock-id`),
}));

// @TODO[PRIO=2]: [Prepare fixtures for components testing].
describe(`Document name update works when`, () => {
  const validPayload: UpdateDocumentNamePayload = {
    name: {
      raw: `Test Document For Me`,
      slug: `test-document-for-me`,
      path: `/test-document-for-me/`,
      segments: [`test`, `document`, `for`, `me`],
    },
    id: `document-id`,
    mdate: `2024-01-01T00:00:00Z`,
  };
  const projectId = `project-id`;
  const uid = `user-uid`;

  describe(`throws an error`, () => {
    it(`if there is no documents`, async () => {
      try {
        await updateDocumentNameHandler({
          payload: validPayload,
          context: {
            uid,
            db: {
              collection: () => ({
                doc: () => ({
                  get: async () => ({
                    data: () => undefined,
                  }),
                }),
              }),
            } as any,
            projectId,
          },
        });
      } catch (error: unknown) {
        expect(error).toMatchSnapshot();
      }
    });

    it(`if cannot find document`, async () => {
      try {
        await updateDocumentNameHandler({
          payload: validPayload,
          context: {
            uid,
            db: {
              collection: () => ({
                doc: () => ({
                  get: async () => ({
                    data: (): DocumentsModel => ({
                      'not-existing-document': {
                        cdate: `2024-01-01T00:00:00Z`,
                        mdate: `2024-01-01T00:00:00Z`,
                        name: `Example document name`,
                        code: ``,
                        path: `/example-document-name/`,
                        visibility: DocumentModelVisibility.Private,
                      },
                    }),
                  }),
                }),
              }),
            } as any,
            projectId,
          },
        });
      } catch (error: unknown) {
        expect(error).toMatchSnapshot();
      }
    });

    it(`if there is out of date change attempt`, async () => {
      try {
        await updateDocumentNameHandler({
          payload: validPayload,
          context: {
            uid,
            db: {
              collection: () => ({
                doc: () => ({
                  get: async () => ({
                    data: (): DocumentsModel => ({
                      [validPayload.id]: {
                        cdate: `2024-01-01T00:00:00Z`,
                        mdate: `2024-04-01T00:00:00Z`,
                        name: `Example document name`,
                        code: ``,
                        path: `example-document-name`,
                        visibility: DocumentModelVisibility.Private,
                      },
                    }),
                  }),
                }),
              }),
            } as any,
            projectId,
          },
        });
      } catch (error: unknown) {
        expect(error).toMatchSnapshot();
      }
    });

    it(`if contains duplicate in own documents`, async () => {
      try {
        await updateDocumentNameHandler({
          payload: validPayload,
          context: {
            uid,
            db: {
              collection: () => ({
                doc: () => ({
                  get: async () => ({
                    data: (): DocumentsModel => ({
                      [validPayload.id]: {
                        cdate: `2024-01-01T00:00:00Z`,
                        mdate: `2024-01-01T00:00:00Z`,
                        name: `Test Document For Me-2`,
                        code: ``,
                        path: `/test-document-for-me-2/`,
                        visibility: DocumentModelVisibility.Private,
                      },
                      'other-document-id': {
                        cdate: `2024-01-02T00:00:00Z`,
                        mdate: `2024-01-02T00:00:00Z`,
                        name: `Test Document For Me`,
                        code: ``,
                        path: `/test-document-for-me/`,
                        visibility: DocumentModelVisibility.Private,
                      },
                    }),
                  }),
                }),
              }),
            } as any,
            projectId,
          },
        });
      } catch (error: unknown) {
        expect(error).toMatchSnapshot();
      }
    });

    it(`if validation for permanent document name fails`, async () => {
      try {
        await updateDocumentNameHandler({
          payload: {
            ...validPayload,
            name: {
              raw: `Too Short`,
              slug: `too-short`,
              segments: [`too`, `short`],
              path: `/too-short/`,
            },
          },
          context: {
            uid,
            db: {
              collection: () => ({
                doc: () => ({
                  get: async () => ({
                    data: (): DocumentsModel => ({
                      [validPayload.id]: {
                        cdate: `2024-01-01T00:00:00Z`,
                        mdate: `2024-01-01T00:00:00Z`,
                        name: `Test Document With Different Name`,
                        code: ``,
                        path: `/test-document-with-different-name/`,
                        visibility: DocumentModelVisibility.Permanent,
                        description: `Some description of my document`,
                        tags: [`programming`],
                      },
                    }),
                  }),
                }),
              }),
            } as any,
            projectId,
          },
        });
      } catch (error: unknown) {
        expect(error).toMatchSnapshot();
      }
    });

    it(`if contains duplicate in permanent documents between different authors`, async () => {
      try {
        await updateDocumentNameHandler({
          payload: validPayload,
          context: {
            uid,
            db: {
              collection: () => ({
                get: () => ({
                  docs: [
                    {
                      data: (): DocumentsModel => ({
                        'other-document-id': {
                          cdate: `2024-01-01T00:00:00Z`,
                          mdate: `2024-01-01T00:00:00Z`,
                          name: `Test Document For Me`,
                          code: ``,
                          path: `/test-document-for-me/`,
                          visibility: DocumentModelVisibility.Permanent,
                          description: `Some description of my document`,
                          tags: [`programming`],
                        },
                      }),
                    },
                  ],
                }),
                doc: () => ({
                  get: async () => ({
                    data: (): DocumentsModel => ({
                      [validPayload.id]: {
                        cdate: `2024-01-01T00:00:00Z`,
                        mdate: `2024-01-01T00:00:00Z`,
                        name: `Test Document For Me`,
                        code: ``,
                        path: `/test-document-for-me/`,
                        visibility: DocumentModelVisibility.Permanent,
                        description: `Some description of my document`,
                        tags: [`programming`],
                      },
                    }),
                  }),
                }),
              }),
            } as any,
            projectId,
          },
        });
      } catch (error: unknown) {
        expect(error).toMatchSnapshot();
      }
    });
  });

  it(`updates documents model and returns dto`, async () => {
    const updateSpy = jest.fn();

    const documentModel: DocumentModel = {
      cdate: `2024-01-01T00:00:00Z`,
      mdate: `2024-01-01T00:00:00Z`,
      name: `Test Document For Me`,
      code: ``,
      path: `/test-document-for-me/`,
      visibility: DocumentModelVisibility.Private,
    };

    const dto = await updateDocumentNameHandler({
      payload: validPayload,
      context: {
        uid,
        db: {
          collection: () => ({
            doc: () => ({
              get: async () => ({
                data: (): DocumentsModel => ({
                  [validPayload.id]: documentModel,
                }),
              }),
              update: updateSpy,
            }),
          }),
        } as any,
        projectId,
      },
    });

    expect(updateSpy).toHaveBeenCalledTimes(1);
    expect(updateSpy).toHaveBeenCalledWith({
      [validPayload.id]: {
        ...documentModel,
        path: validPayload.name.path,
        name: validPayload.name.raw,
        mdate: stampMock,
      },
    });
    expect(dto).toEqual({ mdate: stampMock, name: validPayload.name.raw });
  });
});
