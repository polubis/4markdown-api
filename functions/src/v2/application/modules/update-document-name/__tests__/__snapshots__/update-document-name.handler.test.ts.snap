// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Document name update works when throws an error if cannot find document 1`] = `[Error: {"symbol":"not-found","content":"Document not found","message":"Document not found"}]`;

exports[`Document name update works when throws an error if contains duplicate in own documents 1`] = `[Error: {"symbol":"already-exists","content":"Document with provided name already exists, please change name","message":"Document with provided name already exists, please change name"}]`;

exports[`Document name update works when throws an error if contains duplicate in permanent documents between different authors 1`] = `[Error: {"symbol":"already-exists","content":"Document with provided name already exists, please change name","message":"Document with provided name already exists, please change name"}]`;

exports[`Document name update works when throws an error if there is no documents 1`] = `[Error: {"symbol":"not-found","content":"Document data not found","message":"Document data not found"}]`;

exports[`Document name update works when throws an error if there is out of date change attempt 1`] = `[Error: {"symbol":"out-of-date","content":"The document has been already changed","message":"The document has been already changed"}]`;
